src/middlewared/middlewared/utils/size.py

```diff
@@ -1,7 +1,48 @@
 import humanfriendly
+import re
+
 
 MB = 1048576
+RE_SIZE = re.compile(r'^(\d+(?:\.\d+)?)\s*([KMGTPE]?[iI]?[B]?)$', re.IGNORECASE)
 
 
 def format_size(size):
     return humanfriendly.format_size(size, binary=True)
+
+
+def normalize_size(size: str | float | int | None, raise_exception: bool = True) -> int | None:
+    """
+    Convert a size string (e.g., '2 GiB', '500 MiB', '100K', '10G') to bytes.
+    Returns None if size is not set.
+    """
+    if isinstance(size, str) is False:
+        return size
+
+    size = size.strip().upper()
+    match = RE_SIZE.match(size)
+    if not match:
+        if raise_exception:
+            raise ValueError(f'Invalid size format: {size}')
+        return None
+
+    size_value, unit = match.groups()
+    size_value = float(size_value)
+
+    normalized_unit = unit.upper().replace('IB', 'I').replace('B', '')
+
+    unit_multipliers = {
```

Comment:
Since this lookup table is unlikely to change we can define it outside of scope of method.

Reply:
Agree, @Qubad786 please review my DM for an improved version of this function.